rust/src/message/payload/update.rs

```diff
@@ -0,0 +1,432 @@
+/// let mut buffer = UpdateBuffer::new_unchecked(&mut writer);
+///
+/// buffer
+///    .sum_signature_mut()
+///    .copy_from_slice(sum_signature.as_slice());
+/// buffer
+///    .update_signature_mut()
+///    .copy_from_slice(update_signature.as_slice());
+/// // it is important to set the length first, otherwise, "value_mut()" would panic
+/// buffer
+///    .masked_model_mut()
+///    .set_length(8);
+/// buffer
+///    .masked_model_mut()
+///    .value_mut()
+///    .copy_from_slice(&masked_model[4..]);
+/// buffer
+///    .local_seed_dict_mut()
+///    .set_length(228);
+/// buffer
+///    .local_seed_dict_mut()
+///    .value_mut()
+///    .copy_from_slice(&local_seed_dict[4..]);
+/// assert_eq!(writer, bytes);
+/// ```
+pub struct UpdateBuffer<T> {
+    inner: T,
+}
+
+impl<T: AsRef<[u8]>> UpdateBuffer<T> {
+    /// Perform bound checks on `bytes` to ensure its fields can be
+    /// access without panicking, and return an `UpdateBuffer`.
+    pub fn new(bytes: T) -> Result<Self, DecodeError> {
+        let buffer = Self { inner: bytes };
+        buffer
+            .check_buffer_length()
+            .context("invalid UpdateBuffer")?;
+        Ok(buffer)
+    }
+
+    /// Return an `UpdateBuffer` without performing any bound
+    /// check. This means accessing the various fields may panic if
+    /// the data is invalid.
+    pub fn new_unchecked(bytes: T) -> Self {
+        Self { inner: bytes }
+    }
+
+    /// Perform bound checks to ensure the fields can be accessed
+    /// without panicking.
+    pub fn check_buffer_length(&self) -> Result<(), DecodeError> {
+        let len = self.inner.as_ref().len();
+        // First, check the fixed size portion of the
+        // header. UPDATE_SIGNATURE_RANGE is the last field
+        if len < UPDATE_SIGNATURE_RANGE.end {
+            return Err(anyhow!(
+                "invalid buffer length: {} < {}",
+                len,
+                UPDATE_SIGNATURE_RANGE.end
+            ));
+        }
+
+        // Check the length of the length of the masked model field
+        let _ = LengthValueBuffer::new(&self.inner.as_ref()[self.masked_model_offset()..])
+            .context("invalid masked model field length")?;
+
+        // Check the length of the local seed dictionary field
+        let _ = LengthValueBuffer::new(&self.inner.as_ref()[self.local_seed_dict_offset()..])
+            .context("invalid local seed dictionary length")?;
+
+        Ok(())
+    }
+
+    /// Get the offset of the masked model field
+    fn masked_model_offset(&self) -> usize {
+        UPDATE_SIGNATURE_RANGE.end
+    }
+
+    /// Get the offset of the local seed dictionary field
+    fn local_seed_dict_offset(&self) -> usize {
+        let masked_model =
+            LengthValueBuffer::new_unchecked(&self.inner.as_ref()[self.masked_model_offset()..]);
+        self.masked_model_offset() + masked_model.length() as usize
+    }
+}
+
+impl<'a, T: AsRef<[u8]> + ?Sized> UpdateBuffer<&'a T> {
```

Comment:
it would be nice to have the same order for the method definitions across the various impls as in the buffer to help readability.